keps/sig-api-machinery/4355-coordinated-leader-election/README.md

```diff
@@ -0,0 +1,1190 @@
+  point in time when this was true)
+- controller A elects X
+- controller B marks the leader lease as ""End of term" since it believes Y
+  should be leader
+- controller B elects Y as leader
+- controller A marks the leader lease as ""End of term" since it believes X
+  should be leader
+- ...
+
+This can be avoided by tracking resourceVersion or generation numbers of
+resources used to make a decision in the lease being reconciled and authoring
+the controllers to not to write to a lease when the data used is stale compared
+to the already tracked resourceVersion or generation numbers.
+
+One drawback to this approach is that updating the leader election controller
+can cause undefined behavior when multiple instances of the leader election
+controller are "collaborating". It is difficult to test and prove edge cases
+when an update to the leader election controller code is necessary and could
+fight with the previous version during an mixed version state.
+
+#### Running the coordinated leader election controller in KCM
+
+Since the coordinated leader election controller is a controller that is
+elected, it would also make sense to run in KCM. However, a major drawback is
+that KCM forcefully shuts down when it loses the leader lock and it is possible
+that the leader election controller on the same KCM instance is the leader at
+that time. This causes the coordinated leader election controller to change
+leaders which could cause disruptions.
+
+Two ways to solve this are to gracefully shutdown the KCM and fork the process
+such that the coordinated leader election controller is unaffected. Gracefully
+shutting down the KCM is difficult as controllers are used to the KCM forcefully
+shutting them, and we have no guarantee that third party controllers do not rely
+on this "feature". Forking the process causes additional overhead that we'd like
+to avoid.
+
+#### Running the coordinated leader election controller in a new container
+
+Instead of running in KCM, the coordinated leader election controller could be
+run in a new container (eg: `kube-coordinated-leader-election`). There will be a
+large memory footprint with this approach and adding a new component to the
+control plane could change our Kubernetes architecture in an undesirable way.
+
+### Component instances pick a leader without a coordinator
+
+- A candidates is picked at random to be an election coordinator, and the
+   coordinator picks the leader:
+  - Components race to claim the lease
+  - If a component claims the lease, the first thing it does is check the
+    identity leases to see if there is a better leader
+  - If it finds a better lease, it assigns the lease to that component instead
+    of itself
+
+Pros: 
+  - No coordinated election controller
+
+Cons: 
+
+  - All component instances must watch the identity leases
+  - All components must have the code to decide which component is the best
+    leader
```

Comment:
do we think about more than 3 kcms and 3 ks? "All component" sounds like many. Why is kcm and ks watching some resource bad?

Reply:
addon manager and custom leader elected components is what comes to mind. Agree this should not be "all components". It's not bad, just adds additional complexity over the current best approach.